src/server/game/Chat/Channels/Channel.cpp

```diff
@@ -191,36 +191,38 @@ void Channel::JoinChannel(Player* player, std::string const& pass)
 
     player->JoinedChannel(this);
 
-    if (_announce && !player->GetSession()->HasPermission(rbac::RBAC_PERM_SILENTLY_JOIN_CHANNEL))
+    if (_announceEnabled && !player->GetSession()->HasPermission(rbac::RBAC_PERM_SILENTLY_JOIN_CHANNEL))
     {
         WorldPacket data;
         MakeJoined(&data, guid);
         SendToAll(&data);
     }
 
-    PlayerInfo pinfo;
-    pinfo.player = guid;
+    bool newChannel = _playersStore.empty();
+
+    PlayerInfo& pinfo = _playersStore[guid];
     pinfo.flags = MEMBER_FLAG_NONE;
-    playersStore[guid] = pinfo;
+    pinfo.invisible = !player->isGMVisible();
 
     WorldPacket data;
     MakeYouJoined(&data);
     SendToOne(&data, guid);
 
-    JoinNotify(guid);
-
     // Custom channel handling
     if (!IsConstant())
     {
         // Update last_used timestamp in db
-        if (!playersStore.empty())
-            UpdateChannelUseageInDB();
+        UpdateChannelUseageInDB();
 
         // If the channel has no owner yet and ownership is allowed, set the new owner.
-        if (!_ownerGUID && _ownership)
+        // or if the owner was a GM with .gm visible off
+        // don't do this if the new player is, too, an invis GM, unless the channel was empty
```

Comment:
this ".gm visible off" stuff seems unrelated to "Changes proposed:" section of the PR description